Idle protocol

raazvvann opened the pull request:
Add idle protocol

My question is if it's ok how I get notified about input events, because on rootstoon mouse motion on an empty desktop will lead to idle state bacause in this case rootston doesn't call wlr_seat_pointer_notify_motion
  
I am heghe on IRC

emersion commented:
Maybe `wlr_idle_notify_activity(idle, seat)` is better?

raazvvann commented:
@emersion You say create to have `wlr_idle_notify_activity` which will be called from compositor when there are input events? I wanted to stay in wlroots with input events notification, but what you say, if I understood correctly, seems nicer.

emersion commented:
Some events are not sent to the seat (e.g. pointer outside a view as you said, or compositor keybindings) so I'm not sure if it's possible to only listen to the `wlr_seat`.

raazvvann commented:
Ok. I get it. Thanks for the info. I'll finish this tomorrow.

acrisci commented on types/wlr_idle.c:
```diff
@@ -0,0 +1,159 @@
+
+static void idle_timeout_destroy(struct wlr_idle_timeout *timer) {
+	wl_list_remove(&timer->input_listener.link);
+	wl_event_source_timer_update(timer->idle_source, 0);
```
need to call wl_event_source_remove?

acrisci commented:
The protocol seems to intend to specifically tie this to the activity of a seat and not the input devices themselves.

acrisci commented on include/wlr/types/wlr_idle.h:
```diff
@@ -0,0 +1,32 @@
+#define WLR_TYPES_WLR_IDLE_H
+
+#include <wayland-server.h>
+
```
Can you add a bit of documentation to the header to explain what all this does?

acrisci commented:
Are there any applications that use this in the wild that we can test?

emersion commented:
>The protocol seems to intend to specifically tie this to the activity of a seat and not the input devices themselves.

Yeah, but in cases mentioned above, the seat activity isn't forwarded to the `wlr_seat`.

acrisci commented:
But what is "seat activity"? There are multiple ways to interpret this.

1. If you see the seat as a "collection of input devices", you could say that "seat activity" is any activity of an input device associated with the seat.

2. With a stricter reading of the protocol, you could interpret "seat activity" as only when the seat sends input events.

The example use case of the protocol given seems to me to work best with 1). In that case (since the seat doesn't know what devices are attached to it in our input model) would it be better to leave the input notifications entirely up to rootston since it is in charge of associating input devices with the seat, instead of this way which leaves the user in charge of input notification only some of the time?

acrisci commented on types/wlr_idle.c:
```diff
@@ -0,0 +1,159 @@
+static void handle_activity(struct wlr_idle_timeout *timer) {
+	// rearm the timer
+	wl_event_source_timer_update(timer->idle_source, timer->timeout);
+	// in case the previouse state was sleeping send a resume event and switch state
```
typo here

emersion commented:
>leave the input notifications entirely up to rootston

That's exactly what's planned with `wlr_idle_notify_activity(idle, seat)`.

raazvvann commented:
I implemented `wlr_idle_notify_activity(seat)` but I don't send the wlr_idle, because I didn't want to add wlr_idle all over input events, so I use only the wlr_seat to pass on the notification. With this changes the compositor is responsible to send an activity notification when there is input events.

@acrisci AFAIK the kde plasma is using this protocol. But I implemented a simple idle example in which I used all the features of the protocol.

emersion commented on types/wlr_idle.c:
```diff
@@ -0,0 +1,162 @@
+	return idle;
+}
+
+void wlr_idle_notify_activity(struct wlr_seat *seat) {
```
This should be in the seat, and used by rootston.

acrisci commented on rootston/seat.c:
```diff
@@ -14,6 +15,7 @@
 static void handle_keyboard_key(struct wl_listener *listener, void *data) {
 	struct roots_keyboard *keyboard =
 		wl_container_of(listener, keyboard, keyboard_key);
+	wl_signal_emit(&keyboard->seat->seat->events.activity_notify, NULL);
```
I don't like emitting the raw event on the seat signal from rootston. Having the signal on the seat just for the one protocol also doesn't seem right to me. What about notifying the wlr-idle itself from here? Sort of like `wlr_idle_notify_activity(seat)` which then handles the event on its own.

emersion replied:
+1. `wlr_idle_notify_activity(idle, seat)` could iterate over all clients and update them accordingly.

raazvvann replied:
that means that I should have wlr_idle accessible here in rootston seat? this will lead to removing wlr_seat form idle, but that's not a problem.

emersion replied:
You can access idle from rootston via `desktop->idle`.

acrisci commented:
If a seat is destroyed, all of its associated timers should probably be destroyed as well.

raazvvann commented:
The example it's a bit slow because of that `sleep(1)` in the main loop. I could use `wl_display_dispatch`, but then I couldn't close the program when the thread that should close the program if example it's started with the additional -c argument. If it's not ok, I could spent more time on this, but I think it's good enough.

emersion commented on rootston/seat.c:
```diff
@@ -14,6 +15,8 @@
 	struct roots_keyboard *keyboard =
 		wl_container_of(listener, keyboard, keyboard_key);
+	struct roots_desktop *desktop = keyboard->input->server->desktop;
+	wl_signal_emit(&desktop->idle->activity_notify, NULL);
```
Why don't you call `wlr_idle_notify_activity` in rootston?

raazvvann replied:
where else in rootston I could call this?

emersion replied:
No, the place from which you notify activity (ie. adding a line here) is fine. But you're triggering the event manually. Can you call `wlr_idle_notify_activity` instead? (`wlr_idle_notify_activity` is unused otherwise)

raazvvann replied:
You are totally right. My bad. Sorry about this.

emersion commented on include/wlr/types/wlr_idle.h:
```diff
@@ -0,0 +1,48 @@
+ * Send notification to restart all timers. Called by compositor when
+ * there is an user activity event.
+ */
+void wlr_idle_notify_activity(struct wlr_idle *idle);
```
The seat on which the activity needs to be notified is missing.

raazvvann replied:
I no longer need the seat now, because wlr_idle holds the wl_signal to notify the timers on user activity

emersion replied:
You need both the `wlr_idle` and the `wlr_seat`. When a client wants to be notified when a seat is idle, it doesn't care if another seat is not idle. In other words: when I move my mouse, the mouse is tied to a seat, and that should reset the idle timer for this seat only, not for all seats.

emersion commented on types/wlr_idle.c:
```diff
@@ -0,0 +1,177 @@
+	timer->input_listener.notify = handle_input_notification;
+	wl_signal_add(&idle->activity_notify, &timer->input_listener);
+	// create the timer
+	timer->idle_source = wl_event_loop_add_timer(idle->event_loop, idle_notify, timer);
```
Nitpick: add a NULL check here.

emersion commented on types/wlr_idle.c:
```diff
@@ -0,0 +1,177 @@
+	idle->display_destroy.notify = handle_display_destroy;
+	wl_display_add_destroy_listener(display, &idle->display_destroy);
+
+	idle->wl_global = wl_global_create(display, &org_kde_kwin_idle_interface, 1, idle, idle_bind);
```
Nitpick: 80 line length limit

acrisci commented on examples/idle.c:
```diff
@@ -0,0 +1,171 @@
+	};
+
+	if (simulate_activity_timeout != 0 && simulate_activity_timeout < close_timeout) {
+		if(pthread_create(&t1, NULL, &simulate_activity, (void *)&arg) != 0)
```
need a space after `if` (here and above) and always use braces.

acrisci commented on examples/idle.c:
```diff
@@ -0,0 +1,171 @@
+	fprintf(stdout, "waiting\n");
+	while (run) {
+		wl_display_roundtrip(display);
+		sleep(1);
```
This isn't the way to run the main loop. I think you want to use `wl_display_run()`.

raazvvann replied:
`wl_display_run` is part of the wayland server API. This example is suppose to be a wayland client. I'll rework the example to use `wl_display_dispatch`  and get rid of the `sleep`

acrisci commented on examples/idle.c:
```diff
@@ -0,0 +1,171 @@
+	}
+
+	if (simulate_activity_timeout != 0)
+		pthread_join(t1, NULL);
```
braces for all conditional blocks.

acrisci commented on examples/meson.build:
```diff
@@ -32,3 +32,10 @@ executable(
+executable(
+	'idle',
+	'idle.c',
+	dependencies: [wayland_client, wlr_protos, wlroots, dependency('threads')],
```
the 'threads' dependencies should be declared with the other dependencies.

acrisci commented on types/wlr_idle.c:
```diff
@@ -0,0 +1,188 @@
+		wl_list_remove(&timer->input_listener.link);
+		wl_list_remove(&timer->seat_destroy.link);
+		wl_resource_set_user_data(timer->resource, NULL);
+		free(timer);
```
post no memory to the client.

acrisci commented on types/wlr_idle.c:
```diff
@@ -0,0 +1,188 @@
+	idle->wl_global = wl_global_create(display, &org_kde_kwin_idle_interface,
+				1, idle, idle_bind);
+	if (idle->wl_global == NULL){
+		free(idle);
```
need to do the full cleanup here.

emersion commented:
@acrisci It seems your comments have been addressed, are you okay to merge this?

acrisci commented:
👍
